core/src/com/unciv/logic/civilization/CapitalConnectionsFinder.kt

```diff
@@ -0,0 +1,104 @@
+    private val harbor = "Harbor"
+
+    init {
+        citiesReachedToMediums[civInfo.getCapital()] = hashSetOf("Start")
+    }
+
+    fun find(): Map<CityInfo, Set<String>> {
+        // We map which cities we've reached, to the mediums they've been reached by -
+        // this is so we know that if we've seen which cities can be connected by port A, and one
+        // of those is city B, then we don't need to check the cities that B can connect to by port,
+        // since we'll get the same cities we got from A, since they're connected to the same sea.
+        while (citiesToCheck.isNotEmpty() && citiesReachedToMediums.size < allCivCities.size) {
+            newCitiesToCheck = mutableListOf()
+            for (cityToConnectFrom in citiesToCheck) {
+                if (cityToConnectFrom.containsHarbor()) {
+                    checkHarbor(cityToConnectFrom)
+                }
+                if (railroadIsResearched) {
+                    checkRailroad(cityToConnectFrom)
+                }
+                if (theWheelIsResearched) {
+                    checkRoad(cityToConnectFrom)
+                }
+            }
+            citiesToCheck = newCitiesToCheck
+        }
+        return citiesReachedToMediums
+    }
+
+    private fun checkRoad(cityToConnectFrom: CityInfo) {
+        check(
+                cityToConnectFrom = cityToConnectFrom,
+                transportType = road,
+                overridingTransportType = railroad,
+                tileFilter = { tile -> tile.hasRoad(civInfo) || tile.hasRailroad() || tile.owningCity?.getCenterTile() == tile },
+                cityFilter = { bfs, city -> bfs.tilesReached.containsKey(city.getCenterTile()) }
+        )
+    }
+
+    private fun checkRailroad(cityToConnectFrom: CityInfo) {
+        check(
+                cityToConnectFrom = cityToConnectFrom,
+                transportType = railroad,
+                tileFilter = { tile -> tile.hasRailroad() || tile.owningCity?.getCenterTile() == tile },
+                cityFilter = { bfs, city -> bfs.tilesReached.containsKey(city.getCenterTile()) }
+        )
+    }
+
+    private fun checkHarbor(cityToConnectFrom: CityInfo) {
+        check(
+                cityToConnectFrom = cityToConnectFrom,
+                transportType = harbor,
+                tileFilter = { tile -> tile.isWater || tile.isCityCenter() },
+                cityFilter = { bfs, city ->
+                    bfs.tilesReached.containsKey(city.getCenterTile())
+                            && city.containsHarbor()
+                }
+        )
+    }
+
+    private fun CityInfo.containsHarbor() =
+            this.cityConstructions.containsBuildingOrEquivalent(harbor)
+
+    private fun check(cityToConnectFrom: CityInfo,
+                      transportType: String,
+                      overridingTransportType: String? = null,
+                      tileFilter: (TileInfo) -> Boolean,
+                      cityFilter: (BFS, CityInfo) -> Boolean) {
```

Comment:
Since *all* usages check if city is reached by BFS, this doesn't need to be part of this filter function - all you need is a filter on the cities themselves